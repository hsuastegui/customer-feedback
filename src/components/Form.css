.form {
    display: block;
    padding: 5px;
}

.form label {
    font-size: 12px;
    margin-bottom: 2px;
    display: block;
    color: #232126;
}

.form input:not([type="range"]),
.form textarea {
    display: block;
    width: 100%;
    border-color: #232126;
    border-radius: 5px;
    border-width: 1px;
    line-height: 26px;
    margin-bottom: 10px;
    padding: 0 5px;
    box-sizing: border-box;
}

.form input[type="range"] {
    display: block;
    width: 100%;
    line-height: 26px;
    margin-bottom: 10px;
}

.form button {
    display: block;
    background: #232126;
    color: #D2D9CC;
    height: 36px;
    border-radius: 5px;
    width: 100%;
    cursor: pointer;
}

.form svg:hover {
    cursor: pointer;
    transform: scale(1.1, 1.1);
}