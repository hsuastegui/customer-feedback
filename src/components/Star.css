svg {
    transition: all ease 0.3s;
    fill: #232126;
}

svg.active {
    fill: #BF5630
}