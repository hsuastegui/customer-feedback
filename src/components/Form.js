import React, { useRef, useState } from "react";
import PropTypes from "prop-types";
import Rating from "./Rating";
import "./Form.css";

function Form({ handleSubmit }) {
  const [rating, setRating] = useState(3);
  const formRef = useRef(null);
  const clearForm = (elements) => {
    setRating(3);
    elements.forEach((e) => {
      e.value = "";
    });
  };
  const onSubmit = (e) => {
    e.preventDefault();
    const elements = formRef.current.querySelectorAll("input, textarea");
    const data = Object.values(elements).reduce(
      (acc, curr) => ({ ...acc, [curr.name]: curr.value }),
      {}
    );
    data.rating = rating;
    handleSubmit(data);
    clearForm(Object.values(elements));
  };
  return (
    <form className="form" ref={formRef} onSubmit={onSubmit}>
      <label htmlFor="name">Name</label>
      <input type="text" name="name" required />
      <label htmlFor="email">Email</label>
      <input type="email" name="email" required />
      <label>Rating</label>
      <Rating rating={rating} setRating={setRating} />
      <label htmlFor="comment">Comment</label>
      <textarea name="comment" required></textarea>
      <button type="submit">Send</button>
    </form>
  );
}

Form.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
};

export default Form;
