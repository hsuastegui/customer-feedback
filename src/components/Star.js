import React from "react";
import PropTypes from "prop-types";
import "./Star.css";

function Star({ active, handleClick, value }) {
  const handleRating = () => {
    handleClick(value);
  };
  return (
    <svg
      width="24"
      height="24"
      viewBox="0 0 24 24"
      focusable="false"
      className={active ? "active" : ""}
      onClick={handleRating}
    >
      <path d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"></path>
    </svg>
  );
}

Star.defaultProps = {
  active: false,
  handleClick: () => {},
  value: 0,
};

Star.propTypes = {
  active: PropTypes.bool,
  handleClick: PropTypes.func,
  value: PropTypes.number,
};

export default Star;
