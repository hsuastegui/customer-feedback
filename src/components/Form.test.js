import React from "react";
import { mount, shallow } from "enzyme";
import Form from "./Form";
import Star from "./Star";

describe("Form", () => {
  const handleSubmit = jest.fn();

  it("renders", () => {
    const wrapper = shallow(<Form handleSubmit={handleSubmit} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("handles submit", () => {
    const name = "abc";
    const email = "abc@mail.com";
    const comment = "Lorem ipsum";
    const wrapper = mount(<Form handleSubmit={handleSubmit} />);
    wrapper.find("input").at(0).instance().value = name;
    wrapper.find("input").at(1).instance().value = email;
    wrapper.find("textarea").at(0).instance().value = comment;
    wrapper.find(Star).at(4).simulate("click");
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    expect(handleSubmit).toHaveBeenCalledWith({
      comment,
      email,
      name,
      rating: 5,
    });
    expect(wrapper.find("input").at(0).instance().value).toEqual("");
    expect(wrapper.find("input").at(1).instance().value).toEqual("");
    expect(wrapper.find("textarea").at(0).instance().value).toEqual("");
  });
});
