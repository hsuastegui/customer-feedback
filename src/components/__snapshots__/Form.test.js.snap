// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    Name
  </label>
  <input
    name="name"
    required={true}
    type="text"
  />
  <label
    htmlFor="email"
  >
    Email
  </label>
  <input
    name="email"
    required={true}
    type="email"
  />
  <label>
    Rating
  </label>
  <Stars
    rating={3}
    setRating={[Function]}
  />
  <label
    htmlFor="comment"
  >
    Comment
  </label>
  <textarea
    name="comment"
    required={true}
  />
  <button
    type="submit"
  >
    Send
  </button>
</form>
`;
